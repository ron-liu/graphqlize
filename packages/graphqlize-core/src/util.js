export {
	not, converge, filter, applySpec, evolve, lensProp, over, toPairs, fromPairs, lensIndex,
	contains, join, when, prop, equals, find, tap, map, path, range, inc, dec,
	pick, mergeAll, omit, mapObjIndexed, or, assoc, props, isNil, type, pathEq, length, drop, forEach, groupWith,
	propEq, pair, values, flatten, merge, flip, useWith, ifElse, reject, init, take, skip, subtract,
	head, toUpper, tail, of, toLower, pipe, curry, objOf, __, propOr, has, uniq, groupBy, either, is, all,
	and, set, lensPath, reduce, view, last, apply, keys, call, curryN, mergeWith, remove, split, both, F, toString,
	gt, lt, gte, lte, zipObj, forEachObjIndexed, pathSatisfies, add, propSatisfies, pickBy, multiply, any, compose, isEmpty
} from 'ramda'

export {
	identity as I,
	constant as K
} from 'folktale/core/lambda'
export {concat} from 'folktale/fantasy-land'

export const isArray = Array.isArray

export const Box = x =>
({
	map: f => Box(f(x)),
	fold: f => f(x),
	inspect: () => `Box(${x})`
})

export const printJson = x => console.log(JSON.stringify(x, null, '\t'))

export const SequelizeJsonType = {
	type: Sequelize.JSON,
	get: function(name) {
		const v = this.getDataValue(name)
		if (typeof v === 'string') {
			try {
				return JSON.parse(v)
			}
			catch (e) {
				return v
			}
		} else {
			return v
		}
	}
}
