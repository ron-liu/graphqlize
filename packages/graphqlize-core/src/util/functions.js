import {curry, prop, pipe, isNil, not, equals} from 'ramda'
export {
	not, converge, filter, applySpec, evolve, lensProp, over, toPairs, fromPairs, lensIndex,
	contains, join, when, prop, equals, find, tap, map, path, range, inc, dec,
	pick, mergeAll, omit, mapObjIndexed, or, assoc, props, isNil, type, pathEq, length, drop, forEach, groupWith,
	propEq, pair, values, flatten, merge, flip, useWith, ifElse, reject, init, take, skip, subtract,
	head, toUpper, tail, of, toLower, pipe, curry, objOf, __, propOr, has, uniq, groupBy, either, is, all,
	and, set, lensPath, reduce, view, last, apply, keys, call, curryN, mergeWith, remove, split, both, F, toString,
	gt, lt, gte, lte, zipObj, forEachObjIndexed, pathSatisfies, add, propSatisfies, pickBy, multiply, any, compose, isEmpty, concat,
	identity as I,
	always as K
} from 'ramda'

export const isArray = Array.isArray

export const propFn = curry((name, fn, obj) => fn(prop(name, obj)))
export const isNotNil = pipe(isNil, not)
export const notEquals = curry( (a, b) => pipe(equals(a), not)(b) )
