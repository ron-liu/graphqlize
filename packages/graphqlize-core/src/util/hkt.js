import {
	of as taskOf,
	rejected as taskRejected,
	do as taskDo,
	task,
	waitAll as taskAll,
	fromPromised as taskifyPromiseFn
} from 'folktale/concurrency/task'

export const Box = x =>
	({
		map: f => Box(f(x)),
		ap: g => g.map(x),
		fold: f => f(x),
		inspect: () => `Box(${x})`
	})

export const resultToTask = x => x.fold(taskRejected, taskOf)

export const validationToTask = resultToTask

export const taskTry = fn => {
	try {
		return taskOf(fn())
	}
	catch (e) {
		return taskRejected(e)
	}
}

export const promiseToTask = promise => task(
	({resolve, reject}) => promise.then(resolve).catch(reject)
)

export {taskOf, taskRejected, taskDo, task, taskAll, taskifyPromiseFn}
