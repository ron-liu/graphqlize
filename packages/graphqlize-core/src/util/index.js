export * from './functions'
export * from './hkt'
export * from './misc'
