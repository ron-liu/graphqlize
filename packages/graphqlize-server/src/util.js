import {task} from 'folktale/concurrency/task'
export const promiseToTask = promise => task(
	({resolve, reject}) => promise.then(resolve).catch(reject)
)