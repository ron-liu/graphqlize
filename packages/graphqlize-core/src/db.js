import Sequelize from 'sequelize'
import { GraphqlizeOption, Connection, Db} from './types'
import {Box, pipe, props, K, curry, prop, isNil, promiseToTask, Task, map, ifElse, tap, path, taskOf, taskTry} from './util'
import {CurriedFn2, Fn1} from './basic-types'

// GraphqlizeOption -> Task error db
export const initSequelize = option => taskTry(
	() => Box(option)
		.map(prop('connection'))
		.map(props([ 'database', 'username', 'password', 'option' ]))
		.fold(x => new Sequelize(...x))
)

// GraphqlizeOption -> db -> Task
export const registerGetDbService = (option, db) => taskTry(
	() => Box(option)
		.map(prop('core'))
		.fold(core => core.buildAndAddService({
			name: 'getDb',
			func: ({}, _) => db
		}))
)

// GraphqlizeOption -> Db -> Task
export const sync = curry((option, connector) => pipe(
	ifElse(
		pipe(path(['connection', 'option', 'sync']), isNil),
		taskOf,
		pipe(path(['connection', 'option', 'sync']), x => connector.sync(x), promiseToTask)
	),
	map(K(true))
)(option))


