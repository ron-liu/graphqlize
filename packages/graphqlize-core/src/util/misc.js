import Sequelize from 'sequelize'
import {converge, pipe, head, toUpper, tail, concat, join, curry} from './functions'
import {toLower} from "ramda";
import type {CurriedFn3} from "../basic-types";

export const printJson = x => console.log(JSON.stringify(x, null, '\t'))

export const SequelizeJsonType = {
	type: Sequelize.JSON,
	get: function(name) {
		const v = this.getDataValue(name)
		if (typeof v === 'string') {
			try {
				return JSON.parse(v)
			}
			catch (e) {
				return v
			}
		} else {
			return v
		}
	}
}

export const capitalize: (word: string) => string = converge(
	concat, [
		pipe(head, toUpper),
		tail
	]
)

export const deCapitalize = converge(concat, [ pipe(head, toLower), tail ])

export const joinGraphqlItems = join(', ')

export const surround:CurriedFn3<string, string, string, string>
= curry((xs, before, after) => `${before}${xs}${after}`)
