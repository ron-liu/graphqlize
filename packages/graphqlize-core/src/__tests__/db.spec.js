import {getModelsFromTypes, createSequelize} from './shared'
import {defineSequelizeModels, sync} from '../db'
import {promiseToTask} from "../util/hkt";

test('isUnique should work', async () => {
	const types = [`
	type Person {
		id: ID
		name: String @isUnique
	}
	`]
	const models = await getModelsFromTypes(types)
	const sequelize = createSequelize()
	
	await defineSequelizeModels(sequelize, models)
		.chain(() => sync({connection: {option: {sync: {force: true}}}}, sequelize))
		.run()
		.promise()
	await sequelize.model('Person').create({name: 'ron'})
	await expect(sequelize.model('Person').create({name: 'ron'}))
		.rejects
		.toMatchSnapshot()
})