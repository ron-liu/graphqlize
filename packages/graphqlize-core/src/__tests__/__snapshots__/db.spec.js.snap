// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isUnique should work 1`] = `[SequelizeUniqueConstraintError: Validation error]`;
